discovery/store.go

```diff
@@ -0,0 +1,111 @@
+package discovery
+
+import (
+	"encoding/binary"
+
+	"io"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// waitingProofKey is the proof key which uniquely identifies the waiting
+// proof object. The goal of this key is distinguish the local and remote
+// proof for the same channel id.
+type waitingProofKey [9]byte
+
+// waitingProof is the storable object, which encapsulate the half proof and
+// the information about from which side this proof came. This structure is
+// needed to make channel proof exchange persistent, so that after client
+// restart we may receive remote/local half proof and process it.
+type waitingProof struct {
+	isRemote bool
+	msg      *lnwire.AnnounceSignatures
+}
+
+// newWaitingProof constructs a new waiting prof instance.
+func newWaitingProof(isRemote bool, proof *lnwire.AnnounceSignatures) *waitingProof {
+	return &waitingProof{
+		isRemote: isRemote,
+		msg:      proof,
+	}
+}
+
+// key returns the key which uniquely identifies waiting proof.
+func (p *waitingProof) key() waitingProofKey {
+	var key [9]byte
+	binary.BigEndian.PutUint64(key[:8], p.msg.ShortChannelID.ToUint64())
+
+	if p.isRemote {
+		key[8] = 1
+	}
+	return key
+}
+
+// oppositeKey returns the key which uniquely identifies opposite waiting proof.
+func (p *waitingProof) oppositeKey() waitingProofKey {
+	var key [9]byte
+	binary.BigEndian.PutUint64(key[:8], p.msg.ShortChannelID.ToUint64())
+
+	if !p.isRemote {
+		key[8] = 1
+	}
+	return key
+}
+
+// Encode writes the internal representation of waiting proof in byte stream.
+// NOTE: Part of the channeldb.Storable interface.
+func (p *waitingProof) Encode(w io.Writer) error {
+	var b [1]byte
+	if p.isRemote {
+		b[0] = 1
+	}
+
+	if _, err := w.Write(b[:]); err != nil {
+		return err
+	}
+
+	if _, err := lnwire.WriteMessage(w, p.msg, 0); err != nil {
```

Comment:
I don't think we actually need to use `WriteMessage` here, as that will include the message type header, which in this case, is already known. Instead the `Encode` method of the `lnwire.AnnouncmentSignature` can be used directly.